sys/net/application_layer/telnet/stdio_telnet.c

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       STDIO over Telnet implementation
+ *
+ * This file implements STDIO via a Telnet server with fallback UART output
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <errno.h>
+
+#include "board.h"
+#include "kernel_defines.h"
+#include "net/telnet.h"
+#if IS_USED(MODULE_PERIPH_UART)
+#include "stdio_uart.h"
+#include "periph/uart.h"
+#endif
+#if IS_USED(MODULE_VFS)
+#include "vfs.h"
+#endif
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static inline void _init_fallback(void)
+{
+#if defined(STDIO_UART_DEV) && IS_USED(MODULE_PERIPH_UART)
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, NULL, NULL);
+#endif
+}
+
+static inline int _write_fallback(const void* buffer, size_t len)
+{
+#if defined(STDIO_UART_DEV) && IS_USED(MODULE_PERIPH_UART)
+    uart_write(STDIO_UART_DEV, buffer, len);
+#else
+    (void)buffer;
+#endif
+
+    return len;
+}
+
+void stdio_init(void)
+{
+    _init_fallback();
+
+#if IS_USED(MODULE_VFS)
+    vfs_bind_stdio();
+#endif
+}
```

Comment:
Seems a bit like this could benefit from https://github.com/RIOT-OS/RIOT/issues/13469 as well :-).

Reply:
certainly!